mflike/foreground.py

```diff
@@ -482,36 +640,274 @@ def _bandpass_construction(self, _initialize=False, **params):
+                            ratioP = _cmb2bb(nub)[..., np.newaxis] * blP / tranb_normP
+                            self.bandint_freqs_P.append([nub, ratioP])
+               
+               # in case we don't want to do band integration, e.g. when we have multifreq bandpass in sacc file
                 if self.bandint_nsteps == 1:
                     nub = fr + shift
                     data_are_monofreq = True
-                    self.bandint_freqs.append(nub)
+                    if "tt" in self.requested_cls or "te" in self.requested_cls:
+                        self.bandint_freqs_T.append(nub)
+                    if "te" in self.requested_cls or "ee" in self.requested_cls:
+                        self.bandint_freqs_P.append(nub)
             # using the bandpass from sacc file
             else:
                 nub = nu_ghz + shift
                 if len(bp) == 1:
                     # Monofrequency channel
                     data_are_monofreq = True
-                    self.bandint_freqs.append(nub[0])
+                    if "tt" in self.requested_cls or "te" in self.requested_cls:
+                        self.bandint_freqs_T.append(nub[0])
+                    if "te" in self.requested_cls or "ee" in self.requested_cls:
+                        self.bandint_freqs_P.append(nub[0])
                 else:
-                    trans_norm = trapezoid(bp * _cmb2bb(nub), nub)
-                    trans = bp / trans_norm * _cmb2bb(nub)
-                    self.bandint_freqs.append([nub, trans])
+                    if not self.use_beam_profile:
+                        trans_norm = trapezoid(bp * _cmb2bb(nub), nub)
+                        trans = bp / trans_norm * _cmb2bb(nub)
+                        if "tt" in self.requested_cls or "te" in self.requested_cls:
+                            self.bandint_freqs_T.append([nub, trans])
+                        if "te" in self.requested_cls or "ee" in self.requested_cls:
+                            self.bandint_freqs_P.append([nub, trans])
+                    else:
+                        if self.bandsh_beams_path:
+                            blT, blP = self.return_beams(exp, nu_ghz, shift)
+                        else:
+                            # not propagating bandpass shifts to the chromatic beams
+                            blT, blP = self.return_beams(exp, nu_ghz, 0.)
+
+                        if "tt" in self.requested_cls or "te" in self.requested_cls:
+                            trans_normT = trapezoid(
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT, nub, axis=0
+                            )
+                            ratioT = (
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT / trans_normT
+                            )
+                            self.bandint_freqs_T.append([nub, ratioT])
```

Comment:
Here we are computing `bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT` 2 times (same for polar). I think we should try to reduce the number of operations especially in part of the code frequently called by the likelihood

```python
if "tt" in self.requested_cls or "te" in self.requested_cls:
  bpT = bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT
  self.bandint_freqs_T.append([nub, bpT / trapezoid(bpT, nub, axis=0)])
if "te" in self.requested_cls or "ee" in self.requested_cls:
  bpP = bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blP
  self.bandint_freqs_P.append([nub, bpP / trapezoid(bpP, nub, axis=0)])
```

Reply:
Done